### ISSUE TITLE
Allow to define the FieldError.message property type

### ISSUE THREAD

**Issue by soullivaneuh** [react-hook-form/react-hook-form#9185]
**Is your feature request related to a problem? Please describe.**
I am currently combining [yup validator]() with react-hook-form and [react-i18next]() following the [yup's official guide]().

The guide recommends to define yup translation locales with objects containing `key` and `values` properties:

```typescript
import { setLocale } from 'yup';

setLocale({
  string: {
    matches: () => ({ key: 'string.matches' })
  },
  number: {
    min: ({ min }) => ({ key: 'number.min', values: { min } }),
    max: ({ max }) => ({ key: 'number.max', values: { max } }),
  },
  mixed: {
    notType: () => ({ key: 'mixed.notType' })
  }
});
```

They using the provided object from using `useForm` with the yup resolver being configured on it:

```tsx
<Controller
  name="priority"
  control={control}
  render={({ field }) => (
    <TextField
      {...field}
      label={t('recordFormInputPriority')}
      error={typeof errors.priority !== 'undefined'}
      helperText={
        typeof errors.priority !== 'undefined'
          ? t(`validation:${errors.priority?.message?.key}`, errors.priority?.message?.values)
          : null
      }
      type="text"
      margin="dense"
      fullWidth
    />
  )}
/>
```

It works, but it produces the following Typescript error compilation:

```
353:69 - error TS2339: Property 'key' does not exist on type 'string'.

353                         ? t(`validation:${errors.priority?.message?.key}`, errors.priority?.message?.values)
                                                                        ~~~
353:102 - error TS2339: Property 'values' does not exist on type 'string'.

353                         ? t(`validation:${errors.priority?.message?.key}`, errors.priority?.message?.values)
                                                                                                         ~~~~~
```

This makes sense as the defined `Message` type is set to `string`.

**Describe the solution you'd like**

Add the possibility to override the typing of the `FieldError.message` property, configurable with `useForm`

In a nutshell, something like this:

```typescript
type FieldErrorMessage = {
  key: string;
  values?: Record<string, string>;
}

const form = useForm<FieldValues, TContext, FieldErrorMessage>
```

**Describe alternatives you've considered**

Currently no alternative found except using `@ts-expect-error` on each errors message usage, which is kinda cumbersome.

**Additional context**

**Comment by bluebill1049** [comment]
thanks for the details feature request, moving it into project board as pending status.

**Comment by vcarel** [comment]
Hi, I just wanted to say that I had another use case. I would like to return a ReactNode instead, so to get bold text, links... anything in my error messages 😀

**Comment by bluebill1049** [comment]
> Hi, I just wanted to say that I had another use case. I would like to return a ReactNode instead, so to get bold text, links... anything in my error messages 😀

hey, @vcarel that's a limitation with TS (without type asset), this is currently supported when using JS.

**Comment by soullivaneuh** [comment]
@bluebill1049 Thanks for considering this feature request, but I see you closed the issue. Is it normal? How can we track the status of it?

**Comment by Moshyfawn** [comment]
> @bluebill1049 Thanks for considering this feature request, but I see you closed the issue. Is it normal? How can we track the status of it?

You feature request has been [pined to our work board](https://github.com/react-hook-form/react-hook-form/projects/1); you can track it there. We don't keep feature requests as open issues but assigning them their progress status instead

**Comment by bluebill1049** [comment]
https://github.com/react-hook-form/react-hook-form/issues
